chainio/txmgr/geometric/geometric_test.go

```diff
@@ -0,0 +1,373 @@
+
+	t.Run("Send n=3 txs sequentially", func(t *testing.T) {
+		n := uint64(3)
+		h := newTestHarness(t)
+
+		for nonce := uint64(0); nonce < n; nonce++ {
+			tx := newUnsignedEthTransferTx(nonce, nil)
+			ctxWithTimeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			_, err := h.txmgr.Send(ctxWithTimeout, tx)
+			require.NoError(t, err)
+		}
+
+	})
+
+	t.Run("Send 2 txs with same nonce", func(t *testing.T) {
+		h := newTestHarness(t)
+
+		unsignedTx := newUnsignedEthTransferTx(0, nil)
+		ctxWithTimeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		_, err := h.txmgr.Send(ctxWithTimeout, unsignedTx)
+		require.NoError(t, err)
+
+		unsignedTx = newUnsignedEthTransferTx(0, nil)
+		_, err = h.txmgr.Send(ctxWithTimeout, unsignedTx)
+		require.Error(t, err)
+	})
+}
+
+func newUnsignedEthTransferTx(nonce uint64, gasFeeCap *big.Int) *types.Transaction {
+	if gasFeeCap == nil {
+		// 1 gwei is anvil's default starting baseFeePerGas on its genesis block
+		gasFeeCap = big.NewInt(1_000_000_000)
+	}
+	return types.NewTx(&types.DynamicFeeTx{
+		ChainID:   chainId,
+		Nonce:     nonce,
+		GasTipCap: big.NewInt(1),
+		GasFeeCap: gasFeeCap,
+		Gas:       21000,
+		To:        testutils.ZeroAddress(),
+		Value:     big.NewInt(1),
+	})
+}
+
+type fakeEthBackend struct {
+	// congestedBlocks can be set to a positive number to start a congestion period.
+	// while congestedBlocks > 0, gasTipCap will increase everytime it is read, simulating a congested network
+	// this way whenever the txmgr reads the gasTipCap to bump a tx and resend it, the gasTipCap will increase
+	// so as to prevent the tx from being mined
+	// everytime a block is mined, congestedBlocks will decrease by 1
+	congestedBlocks uint64
+	gasTipCap       *big.Int
+	baseFeePerGas   *big.Int
+	// mu protects all the below fields which are updated in "mining" goroutines (see Send)
+	mu          sync.Mutex
+	blockNumber uint64
+	// we use a single nonce for now because the txmgr only supports a single sender
+	nonce    uint64
+	mempool  map[uint64]*types.Transaction // nonce -> tx
+	minedTxs map[common.Hash]*types.Receipt
+	logger   logging.Logger
+}
+
+func NewFakeEthBackend() *fakeEthBackend {
+	logger := testutils.NewTestLogger().With("component", "fakeEthBackend")
+	backend := &fakeEthBackend{
+		congestedBlocks: 0,
+		gasTipCap:       big.NewInt(1),             // 1 wei, same default as anvil
+		baseFeePerGas:   big.NewInt(1_000_000_000), // 1 gwei, same default as anvil
+		mu:              sync.Mutex{},
+		blockNumber:     0,
+		nonce:           0,
+		mempool:         make(map[uint64]*types.Transaction),
+		minedTxs:        make(map[common.Hash]*types.Receipt),
+		logger:          logger,
+	}
+	backend.startMining()
+	return backend
+}
+
+func (s *fakeEthBackend) WithBaseFeePerGas(baseFeePerGas *big.Int) *fakeEthBackend {
+	s.baseFeePerGas = baseFeePerGas
+	return s
+}
+
+func (s *fakeEthBackend) startMining() {
```

Comment:
nice! this is one of the reason I like and hate fakes :P

Reply:
what's the reason haha? what's wrong with `startMining`? :P

Reply:
lol I was just mentioning in general about fakes.
love - because it exactly tells you what you need and gives you thta
hate - you have to write it